### DISCUSSION TITLE
How do you escape characters in a lua-match query?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [nvim-treesitter/nvim-treesitter#4939] [Q&A]
Context: I used `:TSInstall comment` to add comment highlighting and parsing.

I'd like comments in Python, for example, to light up if they end in a `!`, like this. And I'd like certain others lines to highlight differently if they end in `?` (for example)

```python
# This is a normal line
# But this line is important!
# Note: Maybe we should ask X person about the code below?
super_old_function()
```

I tried adding my own queries to do this, such as 

```scm
("text" @text.danger @nospell
 (#lua-match? @text.danger "^.+\!$"))

("text" @text.note.question @nospell
 (#lua-match? @text.note.rare "^.+\?$"))
```

I've tried other varieties such as \\!, [!], and anything else that I could think of to try to escape the `!` and `?` characters, nothing worked. I read through `:help treesitter-predicate-lua-match?` and the other areas but didn't see anything on this topic. Any advice on how to escape `!` and `?` to get the desired effect would be appreciated.

**Comment by clason** [comment]
`:h lua-patterns`

**Reply by ColinKennedy** [reply]
Your answer is ambiguous so I'm forced to interpret it to mean "I'll teach you how to fish (this answer)". Unfortunately, that's been hard for me to do.

```
Lua intentionally does not support regular expressions, instead it has limited
"patterns" |luaref-patterns| which avoid the performance pitfalls of extended
regex. Lua scripts can also use Vim regex via |vim.regex()|.

These examples use |string.match()| to demonstrate Lua patterns: >lua

    print(string.match("foo123bar123", "%d+"))
    -- 123
    print(string.match("foo123bar123", "[^%d]+"))
    -- foo
    print(string.match("foo123bar123", "[abc]+"))
    -- ba
    print(string.match("foo.bar", "%.bar"))
    -- .bar
```

Implies that "%" is the escape character however

```scm
("text" @text.danger @nospell
 (#lua-match? @text.danger "^.+%!$"))

("text" @text.note.rare @nospell
 (#lua-match? @text.note.rare "^.+%?$"))
```

Both do not work. So then what part of that `:h` message were you trying to bring attention to? The `Lua intentionally does not support regular expressions, instead it has limited "patterns"` part? Because `^` and `$` are supported patterns. You can see them used in the same `comment/highlights.scm` e.g.

```scm
; Issue number (#123)
("text" @number
 (#lua-match? @number "^#[0-9]+$"))
```

And that query above works. And `.` seems to also work, just from some quick experimentation.

Maybe you were trying to point me actually to `:h vim.pesc()` where I would've gotten a direct answer to my question.

`:lua print(vim.pesc("?"))` which returns `%?`. However if `%?` is correctly escaped, it still doesn't explain why

```scm
("text" @text.note.rare @nospell
 (#lua-match? @text.note.rare "^.+%?$"))
```

does not work.

I'd like believe that you had the best of intentions while being indirect in your response but this is causing more confusion for me than anything. I would appreciate it if you could elaborate your response to give an answer to the original question.

`I'd like comments in Python, for example, to light up if they end in a !, like this. And I'd like certain others lines to highlight differently if they end in ? (for example)`.

How would I write a query in order to do this? Thank you for your help in advance.

**Reply by ColinKennedy** [reply]
Worth noting that the expression + escape is working, for certain characters. For example

```scm
("text" @text.note.rare @nospell
 (#lua-match? @text.note.rare "^.+%$"))
```

This does work.

![Screenshot from 2023-06-11 12-48-26](https://github.com/nvim-treesitter/nvim-treesitter/assets/10103049/c6ae423a-9ccd-4579-a32d-cdc58ecc03dd)

But changing to

```scm
("text" @text.note.rare @nospell
 (#lua-match? @text.note.rare "^.+%?"))
```

does not work

![Screenshot from 2023-06-11 12-48-43](https://github.com/nvim-treesitter/nvim-treesitter/assets/10103049/f77b66b6-6e1a-4732-9255-b90143415376)

There's something invisible at play and it's hard to know what it is.
